torch/csrc/autograd/engine.cpp

```diff
@@ -69,6 +73,47 @@ struct ReadyQueue {
   FunctionTask pop();
 };
 
+// Note [Reentrant backwards]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
+// To understand the reentrant backwards problem, we have to notice two
+// aspects of how the autograd engine is implemented today:
+//
+//  1. When you call Engine::execute(), you want to block until autograd
+//  execution finishes so that you can get the final result variables
+//  of the backwards pass.
+//
+//  2. Autograd execution operates by having a *single* worker assigned
+//  for any given task (based on the device the task it will run on),
+//  which is responsible for executing the contents of the task.
```

Comment:
This is also quite misleading. I would say:

The engine operates by having a single worker thread per work queue, and every work queue is pinned to a specific device where the operation is executed.